File: ts/src/financial/backtesting.ts
/**
 * Backtesting engine
 */

import { Portfolio } from './portfolio';
import { Signal, SignalGenerator, SMACrossover } from './signals';
import { sharpeRatioSync, maxDrawdownSync, dailyReturns } from './risk';
import { isWasmReady } from '../wasm-loader';

/**
 * Result of a backtest run
 */
export interface BacktestResult {
  /** Total return percentage */
  totalReturn: number;
  /** Sharpe ratio */
  sharpeRatio: number;
  /** Maximum drawdown */
  maxDrawdown: number;
  /** Win rate (profitable trades / total trades) */
  winRate: number;
  /** Number of trades executed */
  numTrades: number;
  /** Equity curve (portfolio value over time) */
  equityCurve: number[];
  /** List of trades */
  trades: Trade[];
}

/**
 * A single trade record
 */
export interface Trade {
  type: 'BUY' | 'SELL';
  price: number;
  quantity: number;
  timestamp: number;
  pnl?: number;
}

/**
 * Simple buy-and-hold strategy for comparison
 */
export class SimpleStrategy implements SignalGenerator {
  readonly name = 'Buy and Hold';
  private bought = false;

  generate(prices: number[]): Signal {
    if (!this.bought && prices.length > 0) {
      this.bought = true;
      return Signal.Buy;
    }
    return Signal.Hold;
  }

  generateSeries(prices: number[]): Signal[] {
    return prices.map((_, i) => (i === 0 ? Signal.Buy : Signal.Hold));
  }

  reset(): void {
    this.bought = false;
  }
}

/**
 * Run a backtest on historical price data
 */
export function backtest(
  prices: number[],
  strategy: SignalGenerator,
  options: {
    initialCapital?: number;
    positionSize?: number; // Fraction of capital per trade (0-1)
    commission?: number; // Commission per trade
  } = {}
): BacktestResult {
  const { initialCapital = 10000, positionSize = 1.0, commission = 0 } = options;

  const portfolio = new Portfolio(initialCapital);
  const equityCurve: number[] = [];
  const trades: Trade[] = [];
  let position = 0;
  let entryPrice = 0;

  for (let i = 0; i < prices.length; i++) {
    const price = prices[i];
    const signal = strategy.generate(prices.slice(0, i + 1));

    // Execute trades based on signals
    if (signal === Signal.Buy && position === 0) {
      const capital = portfolio.cash * positionSize;
      const quantity = Math.floor((capital - commission) / price);

      if (quantity > 0) {
        portfolio.buy('ASSET', quantity, price);
        position = quantity;
        entryPrice = price;
        trades.push({
          type: 'BUY',
          price,
          quantity,
          timestamp: i,
        });
      }
    } else if (signal === Signal.Sell && position > 0) {
      const pnl = (price - entryPrice) * position - commission;
      portfolio.sell('ASSET', position, price);
      trades.push({
        type: 'SELL',
        price,
        quantity: position,
        timestamp: i,
        pnl,
      });
      position = 0;
    }

    // Record equity
    const currentValue = portfolio.value({ ASSET: price });
    equityCurve.push(currentValue);
  }

  // Close any remaining position at the end
  if (position > 0) {
    const finalPrice = prices[prices.length - 1];
    const pnl = (finalPrice - entryPrice) * position;
    portfolio.sell('ASSET', position, finalPrice);
    trades.push({
      type: 'SELL',
      price: finalPrice,
      quantity: position,
      timestamp: prices.length - 1,
      pnl,
    });
  }

  // Calculate metrics
  const returns = dailyReturns(equityCurve);
  const sellTrades = trades.filter((t) => t.type === 'SELL');
  const winningTrades = sellTrades.filter((t) => (t.pnl ?? 0) > 0);
  const winRate = sellTrades.length > 0 ? winningTrades.length / sellTrades.length : 0;

  // Calculate performance metrics
  // Note: sharpeRatio and maxDrawdown use WASM - ensure initWasm() was called
  let sharpe = 0;
  let mdd = 0;

  if (isWasmReady()) {
    sharpe = sharpeRatioSync(returns);
    mdd = maxDrawdownSync(equityCurve);
  } else {
    // Fallback to simple calculations if WASM not ready
    const meanReturn = returns.length > 0 ? returns.reduce((a, b) => a + b, 0) / returns.length : 0;
    const variance = returns.length > 0 ? returns.reduce((sum, r) => sum + (r - meanReturn) ** 2, 0) / returns.length : 0;
    const stdDev = Math.sqrt(variance);
    sharpe = stdDev !== 0 ? meanReturn / stdDev : 0;

    // Calculate max drawdown
    let peak = equityCurve[0] || 0;
    for (const val of equityCurve) {
      if (val > peak) peak = val;
      const dd = (peak - val) / peak;
      if (dd > mdd) mdd = dd;
    }
  }

  return {
    totalReturn: (equityCurve[equityCurve.length - 1] - initialCapital) / initialCapital,
    sharpeRatio: sharpe,
    maxDrawdown: mdd,
    winRate,
    numTrades: trades.length,
    equityCurve,
    trades,
  };
}

File: crates/financial/financial-core/src/backtesting.rs
//! Backtesting engine implementation.

use financial_spi::{BacktestResult, Backtester, Signal, Trade, Result, FinancialError};

/// Simple backtesting engine.
#[derive(Debug, Clone, Default)]
pub struct SimpleBacktester {
    trades: Vec<Trade>,
    initial_capital: f64,
}

impl SimpleBacktester {
    /// Create a new backtester with initial capital.
    pub fn new(initial_capital: f64) -> Self {
        Self {
            trades: Vec::new(),
            initial_capital,
        }
    }
}

impl Backtester for SimpleBacktester {
    fn run(&self, signals: &[Signal], prices: &[f64]) -> Result<BacktestResult> {
        if signals.len() != prices.len() {
            return Err(FinancialError::BacktestError(
                format!("Signals length ({}) != prices length ({})", signals.len(), prices.len())
            ));
        }

        if prices.is_empty() {
            return Ok(BacktestResult {
                total_return: 0.0,
                sharpe_ratio: 0.0,
                max_drawdown: 0.0,
                win_rate: 0.0,
                num_trades: 0,
                equity_curve: vec![],
            });
        }

        let mut equity = self.initial_capital;
        let mut equity_curve = Vec::with_capacity(prices.len());
        let mut position: Option<(f64, i64)> = None; // (entry_price, entry_time)
        let mut trades: Vec<Trade> = Vec::new();
        let mut returns: Vec<f64> = Vec::new();
        let mut prev_equity = equity;

        for (i, (&signal, &price)) in signals.iter().zip(prices.iter()).enumerate() {
            match signal {
                Signal::Buy if position.is_none() => {
                    position = Some((price, i as i64));
                }
                Signal::Sell if position.is_some() => {
                    let (entry_price, entry_time) = position.take().unwrap();
                    let pnl = price - entry_price;
                    equity += pnl * (self.initial_capital / entry_price);
                    trades.push(Trade {
                        entry_price,
                        exit_price: price,
                        quantity: self.initial_capital / entry_price,
                        pnl,
                        entry_time,
                        exit_time: i as i64,
                    });
                }
                _ => {}
            }

            equity_curve.push(equity);
            if prev_equity > 0.0 {
                returns.push((equity - prev_equity) / prev_equity);
            }
            prev_equity = equity;
        }

        // Calculate metrics
        let total_return = if self.initial_capital > 0.0 {
            (equity - self.initial_capital) / self.initial_capital
        } else {
            0.0
        };

        let sharpe_ratio = calculate_sharpe(&returns, 0.0);
        let max_drawdown = calculate_max_drawdown(&equity_curve);
        let win_rate = if trades.is_empty() {
            0.0
        } else {
            trades.iter().filter(|t| t.pnl > 0.0).count() as f64 / trades.len() as f64
        };

        Ok(BacktestResult {
            total_return,
            sharpe_ratio,
            max_drawdown,
            win_rate,
            num_trades: trades.len(),
            equity_curve,
        })
    }

    fn trades(&self) -> &[Trade] {
        &self.trades
    }
}

fn calculate_sharpe(returns: &[f64], risk_free_rate: f64) -> f64 {
    if returns.is_empty() {
        return 0.0;
    }
    let mean: f64 = returns.iter().sum::<f64>() / returns.len() as f64;
    let excess = mean - risk_free_rate;
    let variance: f64 = returns.iter().map(|r| (r - mean).powi(2)).sum::<f64>() / returns.len() as f64;
    let std_dev = variance.sqrt();
    if std_dev < 1e-10 {
        0.0
    } else {
        excess / std_dev
    }
}

fn calculate_max_drawdown(equity_curve: &[f64]) -> f64 {
    if equity_curve.is_empty() {
        return 0.0;
    }
    let mut max_dd = 0.0;
    let mut peak = equity_curve[0];
    for &value in equity_curve {
        if value > peak {
            peak = value;
        }
        let dd = (peak - value) / peak;
        if dd > max_dd {
            max_dd = dd;
        }
    }
    max_dd
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_simple_backtest() {
        let backtester = SimpleBacktester::new(10000.0);
        let signals = vec![Signal::Buy, Signal::Hold, Signal::Hold, Signal::Sell];
        let prices = vec![100.0, 105.0, 110.0, 115.0];

        let result = backtester.run(&signals, &prices).unwrap();
        assert!(result.total_return > 0.0);
        assert_eq!(result.num_trades, 1);
    }
}
